lib/memo_wise.rb

```diff
@@ -3,6 +3,49 @@
 require "memo_wise/version"
 
 module MemoWise
-  class Error < StandardError; end
-  # Your code goes here...
+  def initialize(*values)
+    @_memo = {}
+    super
+  end
+
+  def self.prepended(target)
+    class << target
+      # Implements memoization for the given method name.
+      #
+      # The method must be a *pure function* of the value attributes, and
+      # therefore referentially transparent. Replaces the method with a new
+      # method of the same name, which memoizes the result of calling the
+      # original method. See module comment above for a usage example.
+      #
+      # @param method_name [Symbol]
+      #   Name of method for which to implement memoization.
+      def memo_wise(method_name)
+        not_memoized_name = "#{method_name}_not_memoized".to_sym
+        alias_method not_memoized_name, method_name
+        private not_memoized_name
+
+        define_method(method_name) do |*args|
+          key = [method_name, args].freeze
+          @_memo.fetch(key) do
+            @_memo[key] = send(not_memoized_name, *args)
+          end
+        end
```

Comment:
For speed, we will need to replace the meta-programming implementation with the code generation implementation, right?